src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -659,6 +694,24 @@ private static bool TryParseExactX(ReadOnlySpan<char> guidString, ref GuidResult
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Note: On x86 / x64, I believe you'll get better codegen if you use the following pattern:

```cs
DecodeByte(char ch1, char ch2, ref int invalidIfNegative)
{
  int h1 = -1;
  if (/* ... */) { h1 = /* ... */; }
  h1 <<= 4;
  
  int h2 = -1;
  if (/* ... */) { h2 = /* ... */; }
  
  int temp = h1 | h2;
  invalidIfNegative |= temp;
  return (byte)temp;
}

int invalidIfNegative = 0;
DecodeByte(/* ... */, ref invalidIfNegative);
DecodeByte(/* ... */, ref invalidIfNegative);
DecodeByte(/* ... */, ref invalidIfNegative);
if (invalidIfNegative < 0) { /* FAIL */ }
```

There are a few reasons for this:

- Setting the initial value to `-1` instead of `0xFF` allows using minified codegen (x86 / x64-specific)
- It eliminates the need for a temp register to hold `h1 | h2` in addition to `h1 << 4 | h2`
- Comparing a value `< 0` or `>= 0` has a smaller codegen footprint than comparing against the const value `0xFF`

Given that this method is marked aggressive-inline and is unrolled, those first two bullets in particular might result in cascading improvements.

Reply:
Thanks!  I'll merge this and then follow-up on your suggestion, submitting another PR for it if it results in improvements.

Reply:
@GrabYourPitchforks, I'm looking at your suggestion now, and I'm confused how it's supposed to work, in particular for invalid characters.  The lookup table contains bytes, with invalid entries as 0xFF, not -1.  Is your suggestion to add a new Int32-array-based lookup table where invalid entries use -1 instead of 0xFF?

Reply:
Ah, a couple well-placed sbyte casts should do the trick.  I'll try it out...

Reply:
Yes, sbyte casts would work. According to Agner (https://www.agner.org/optimize/instruction_tables.pdf), p. 262, _movzx_ (zero-extension) and _movsx_ (sign-extension) from memory to a register have the same cost.